server/apps/bindings.go

```diff
@@ -0,0 +1,65 @@
+package apps
+
+import (
+	"github.com/mattermost/mattermost-plugin-apps/server/api"
+	"github.com/pkg/errors"
+)
+
+func mergeBindings(bb1, bb2 []*api.Binding) []*api.Binding {
+	out := append([]*api.Binding(nil), bb1...)
+
+	for _, b2 := range bb2 {
+		found := false
+		for i, o := range out {
+			if b2.AppID == o.AppID && b2.LocationID == o.LocationID {
+				found = true
+
+				// b2 overrides b1, if b1 and b2 have Bindings, they are merged
+				merged := b2
+				if len(o.Bindings) != 0 && b2.Call == nil {
+					merged.Bindings = mergeBindings(o.Bindings, b2.Bindings)
+				}
+				out[i] = merged
+			}
+		}
+		if !found {
+			out = append(out, b2)
+		}
+	}
+	return out
+}
+
+func setAppID(bb []*api.Binding, appID api.AppID) {
+	for _, b := range bb {
+		b.AppID = appID
+		if len(b.Bindings) != 0 {
+			setAppID(b.Bindings, appID)
+		}
+	}
+}
+
+// This and registry related calls should be RPC calls so they can be reused by other plugins
+func (s *service) GetBindings(cc *api.Context) ([]*api.Binding, error) {
```

Comment:
TODO: Make sure top-level elements can not be overwritten, maybe switch around the defaulting - first one wins, no overrides, only merging children

Reply:
Filed https://mattermost.atlassian.net/browse/MM-30036 to cover separately, in sync with Autocomplete PRs